### PR TITLE
Ensure auth methods are re-fetched after an authentication error.

### LINKED ISSUE #1793: auth.service.ts doesn't refresh authMethods after AUTHENTICATED_ERROR
**Describe the bug**
We've observed for a while that after the backend is restarted while a user is logged-in, reloading the logged-in browser tab results in an "empty" login menu, like this:

image

We've seen this problem with our DSpace-CRIS installation as well as a recent `main` branch of dspace-angular, and with Firefox 91.11.0esr as well as Chromium 103.0.5060.53.

**To Reproduce**
1. Start frontend and backend
2. Open a browser
3. Log into the frontend in the browser window
4. Restart the backend
5. Reload the browser tab
6. Open login menu

**Expected behavior**
The authentication methods are expected to appear.

**Analysis**
I've finally found the time to investigate and it appears that while `AUTHENTICATED_ERROR` fires correctly, all that's done in response is to remove the token from storage (which is why things will work again if you reload the tab twice); there's no working code path that results in a `RetrieveAuthMethodsAction` or a forced page reload.

I say "no working code path" because this code in `trackTokenExpiration` comes quite close to handling this (and may provide the appropriate UI response to a non-expired token that unexpectedly fails to work):

```
  public trackTokenExpiration(): void {
    let token: AuthTokenInfo;
    let currentlyRefreshingToken = false;
    this.store.pipe(select(getAuthenticationToken)).subscribe((authTokenInfo: AuthTokenInfo) => {
      // If new token is undefined an it wasn't previously => Refresh failed
      if (currentlyRefreshingToken && token !== undefined && authTokenInfo === undefined) {
        // Token refresh failed => Error notification => 10 second wait => Page reloads & user logged out
        this.notificationService.error(this.translateService.get('auth.messages.token-refresh-failed'));
        setTimeout(() => this.navigateToRedirectUrl(this.hardRedirectService.getCurrentRoute()), 10000);
        currentlyRefreshingToken = false;
      }
```

Unfortunately, the code above is never called in our case, because the authentication token never assumes a value other than `undefined`. As an experiment, I've modified the `AUTHENTICATED_ERROR` handler to use `null` rather than `undefined`:

```
    case AuthActionTypes.AUTHENTICATED_ERROR:
    case AuthActionTypes.RETRIEVE_AUTHENTICATED_EPERSON_ERROR:
      return Object.assign({}, state, {
        authenticated: false,
        authToken: null,
        error: (action as AuthenticationErrorAction).payload.message,
        loaded: true,
        blocking: false,
        loading: false
      });
```

and adjusted the code in `trackTokenExpiration` accordingly. That appears to work.

**Proposed fixes**
I think the easiest fix would be to trigger a `CheckAuthenticationTokenCookieAction` (which is very different from a `CheckAuthenticationTokenAction`) in response to an `AUTHENTICATED_ERROR`. However, as that action may itself result in an `AUTHENTICATED_ERROR`, we must do so only if we've actually deleted a token from storage, in order to avoid an infinite loop if the backend is actually throwing an error.

That would, however, result in a transparent log-out as observed by the user, which might not be desired.

### PR DESCRIPTION
**Pull request by philipprumpf**
## References
* Fixes #1793.

## Description
Make sure a `CheckAuthenticationTokenAction` is triggered if authentication fails (this happens when the backend is restarted while a non-expired token is in storage). This, in turn, refreshes the auth methods and avoids an empty non-functional authentication menu.

## Instructions for Reviewers
We must be careful not to trigger a `CheckAuthenticationTokenAction` recursively, which is why `hasToken` was added.

List of changes in this PR:
* add `AuthenticationService.hasToken` to check existence of a token
* make `removeToken` call conditional on the existence of a token
* trigger `CheckAuthenticationTokenAction` if authentication fails and there was a valid-looking token

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [X] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [X] My PR doesn't introduce circular dependencies
- [X] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [ ] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).

### CODE REVIEW

File: src/app/core/auth/auth.service.ts [hunk 1]
```diff
@@ -427,7 +427,14 @@ export class AuthService {
...
    */
   public removeToken() {
-    return this.storage.remove(TOKENITEM);
+    this.storage.remove(TOKENITEM);
+  }
+
+  /**
+   * Check authentication token info
+   */
+  public hasToken(): boolean {
```
Comment (tdonohue) [inline src/app/core/auth/auth.service.ts:436]: You could add a simple unit test here which is similar to how `removeToken()` is tested in `auth.service.spec.ts`.  See https://github.com/DSpace/dspace-angular/blob/main/src/app/core/auth/auth.service.spec.ts#L372-L375

**Review by tdonohue — COMMENTED** [review]
@philipprumpf : Thanks!  I've added a brief comment inline about how to add one simple test.  I'm also uncertain of the best way of adding a test for `auth.effects.ts` (I do see there's an `auth.effects.spec.ts` -- I'm just less familiar with its test structure).  So, I'll assign this to @artlowel for review to see if he can review & provide additional advice.

**Comment by philipprumpf** [comment]
Hmm. I've added some basic tests, but the 14.x tests appear to be failing in unrelated tests. Is this expected?

**Comment by artlowel** [comment]
@philipprumpf yes the e2e tests sometimes fail at random. I'll restart them

**Review by atarix83 — CHANGES_REQUESTED** [review]
thanks @philipprumpf  for this PR.

it overall looks good, i've only found an issue that can be easily resolved.
Essentialy after restarting the rest server now the login methods are retrieved correctly but an error is shown in the login form as you can see in the screenshot

Schermata da 2022-09-07 17-28-13

this can be easily resolved by clearing the property `error` of the auth state when the action RETRIEVE_AUTH_METHODS is dispatched

**Comment by philipprumpf** [comment]
@atarix83 Thanks! That makes sense, yes, I'll make the change.

**Comment by tdonohue** [comment]
@philipprumpf : Just a friendly reminder that this PR is on your plate to do minor updates.  If you'd like this included in 7.4, we'd need those updates done ideally in the next week... any PR not merged by the end of Sept will be rescheduled for the next release (7.5 due in Feb).  So, the quicker you can make updates, the quicker we can get this re-reviewed and ready to go!

**Comment by philipprumpf** [comment]
@tdonohue Thanks!

@atarix83 Thanks again for the review. I've updated the PR.

**Review by atarix83 — APPROVED** [review]
thanks @philipprumpf  LGTM

**Review by tdonohue — CHANGES_REQUESTED** [review]
@philipprumpf :  When I try to reproduce the error describe in #1793, I seem to be hitting an infinite loop...the page never reloads.

Here's what I'm trying:
1. Start up the frontend using the latest `main` code with this PR also installed.
2. Login via the UI
3. Restart the backend.  Wait for it to complete restarting
4. Click reload in my browser window (where I had previously logged into the UI).
5. At this point the browser window hangs & never reloads.  If I look at the backend logs (dspace.log), I see that it appears to be encountering an infinite loop... I see the same requests to `/server/api/authn/status` over and over and over again.  It never ends, until I shutdown the UI.

I **cannot** reproduce this infinite loop (browser window hanging) if I use the latest UI code on `main`.  So, it appears to be caused by this PR.   

**Finally,** I've also noticed that when I run the latest `main`, **I can NO LONGER REPRODUCE #1793**.  This makes me wonder if we've fixed this in a different manner (almost by accident) in a different recent PR.  That may mean this particular PR can just be closed.

**Comment by philipprumpf** [comment]
I agree this particular PR addresses a problem that no longer happens (or at least happens differently - if I notice it again, I'll open a new PR against 7.4)

**Pull request closed without merging**